Bro integration

iBigQ opened the pull request:
Bro integration for osquery. This adds an InternalRunner to the osquery daemon for communication with Bro. This Runner can update the daemon schedule on behalf of Bro. Also, this integration provides a logger plugin that sends query results to Bro. For more information, please refer to this [repository](https://github.com/bro/bro-osquery) for this [presentation](https://www.bro.org//bro4pros2017/Haas_OSquery_Bro4Pros2017.pdf).

Bro integration can be activated by the command line flag --disable-bro=false. There are basically no changes to osquery except starting the Bro Runner and registering the bro logger.

Please note, the code was checked for functional working only. Currently, there are NO unit tests and no memory leak detection. I would appreciate help for writing tests and check for code performance. 

Since this an early pull request, I dont expect this to be merged right now. Instead, I would like to get your feedback and help at an early stage.

theopolis commented:
@poppyseedplehzr or @jedi22, do you mind taking a look at C++. I will review the CMake/provision scripts.

theopolis commented on osquery/CMakeLists.txt:
```diff
@@ -69,6 +69,10 @@ else()
   ADD_OSQUERY_LINK_CORE("glog")
 endif()
 
+ADD_OSQUERY_LINK_CORE("broker")
```
These lines should do most of the work of the `FindX` CMake libraries. We should remove those files in `cmake/*`.

iBigQ replied:
You are right. I used these `FindX` when testing with shared libraries installed locally. I removed them

theopolis commented:
Try running:
```
make format_master
```

To do a large pass at style formatting for the C/C++ code.

theopolis commented on osquery/bro/BrokerManager.cpp:
```diff
@@ -0,0 +1,312 @@
+#include <osquery/flags.h>
```
All the `.cpp` documents will need the project copyright header.

theopolis replied:
We should follow the:
1. C includes
2. C++ includes
3. Broad project dependencies includes
4. File-specific dependency includes
5. osquery includes
6. osquery local includes

Order of includes, keeps us honest and organized.

theopolis commented on osquery/bro/BrokerManager.cpp:
```diff
@@ -0,0 +1,312 @@
+  this->qm = QueryManager::getInstance();
+}
+
+osquery::Status BrokerManager::setNodeID(const std::string& uid) {
```
You're in the `osquery` namespace so feel free to drop the `osquery::` references.

theopolis commented on osquery/bro/BrokerManager.cpp:
```diff
@@ -0,0 +1,312 @@
+}
+
+osquery::Status BrokerManager::setNodeID(const std::string& uid) {
+  if (this->nodeID.empty()) {
```
The use of `this->` is overkill, feel free to drop it throughout the class implementation code.

theopolis commented on osquery/bro/BrokerManager.cpp:
```diff
@@ -0,0 +1,312 @@
+  return this->groups;
+}
+
+/////////////////////////////////////////////////////////
```
No need for fancy comment-header things.

theopolis commented on osquery/bro/BrokerManager.cpp:
```diff
@@ -0,0 +1,312 @@
+Status BrokerManager::logQueryLogItemToBro(const QueryLogItem& qli) {
+  // Attributes from QueryLogItem
+  std::string queryID = qli.name; // The QueryID
+  //  std::string identifier = qli.identifier; // The HostID
```
If this is old/dead code, please delete.

theopolis commented on osquery/bro/BrokerManager.h:
```diff
@@ -0,0 +1,92 @@
+
+ public:
+  // Get a singleton instance
+  static BrokerManager* getInstance() {
```
We have a few examples of singleton use, check out the `Config` or `RegistryFactory` classes for best-practice.

iBigQ replied:
@theopolis Quick question here. I used `new` to create a singleton because it will be on the heap. If I would do:
```
static BrokerManager& getInstance() {
    static BrokerManager bm;
    return bm;
  }
```
as for the `Config`, it would be placed on the Stack. I had this Problem when running the Code as an extension. There have been two threads, each having a singleton on their Stack.

This also applies e.g. for the Distributed and Bro, since they start another threat. Wouldnt this be a Problem?

theopolis replied:
Our existing use is the Meyer's Singleton pattern, the recommended for C++11+ codebases. This works in our multi-threaded environment too.
> `static` storage duration for a function local means that only one instance of that local exists in the program

http://stackoverflow.com/questions/17712001/how-is-meyers-implementation-of-a-singleton-actually-a-singleton

theopolis commented on osquery/bro/BrokerManager.h:
```diff
@@ -0,0 +1,92 @@
+
+  osquery::Status getTopics(std::vector<std::string>& topics);
+
+  osquery::Status peerEndpoint(std::string ip, int port);
```
Double check that you need this as a copied string and not a `const` reference.

theopolis commented on osquery/bro/BrokerManager.h:
```diff
@@ -0,0 +1,92 @@
+  // The singleton object
+  static BrokerManager* _instance;
+
+  QueryManager* qm = nullptr;
```
Can we use `std::unique_pointer` and `std::shared_pointer` for all pointers?

theopolis commented:
Wow, there's a LOT of code here.

It will take a bit of time to bring this into osquery's style, performance, best-practice, so please be patient haha.

iBigQ commented on osquery/bro/BrokerManager.cpp:
```diff
@@ -0,0 +1,312 @@
+
+  // Common message fields
+  std::string uid = this->getNodeID();
+  std::string topic = this->qm->getEventTopic(queryID);
```
As these are common message fields, they are equal for all messages that will be created in the following `for`.  I think it is better to set them once and use these local variable later for every message that is created

iBigQ commented on osquery/bro/BrokerManager.cpp:
```diff
@@ -0,0 +1,312 @@
+
+Status BrokerManager::logQueryLogItemToBro(const QueryLogItem& qli) {
+  // Attributes from QueryLogItem
+  std::string queryID = qli.name; // The QueryID
```
I added this for better readability and maintenance of the code.

iBigQ commented on osquery/bro/BrokerManager.cpp:
```diff
@@ -0,0 +1,312 @@
+    msg.push_back(result_info);
+
+    // Format each column
+    for (std::tuple<std::string, ColumnType, ColumnOptions> t : columns) {
```
The nested `for`s are relevant for error conditions only. They give some detailed information about the cause. Meanwhile I identified the reason why this error happened to me. I would keep this code in case related errors occur

iBigQ commented:
I updated the RP. Still TODO

- [ ] Singleton and kInstance_
- [ ] unique/shared pointers

iBigQ commented on osquery/bro/BrokerManager.cpp:
```diff
@@ -0,0 +1,309 @@
+namespace osquery {
+
+
+    BrokerManager *BrokerManager::_instance = nullptr;
```
this is necessary to initialize the kInstance_ of Singletons if they are stored as pointers. The Singleton can only be retrieved by Class::getInstance()

theopolis commented:
I'm curious why you're resistant to using smart pointers? :P

> The use of pointers is very limited and there is a well define creation and deletion point in code execution. I dont see any reasons to adapt this to smart pointers.

That's like "well I'm a safe driver, so why should I wear a seat belt?". There are lots of other people contributing to this project, who may need to make modifications to your code. You should set them up for success and ease the burden of memory responsibility if it's easy. :)

theopolis commented on osquery/bro/utils.cpp:
```diff
@@ -0,0 +1,197 @@
+                                 SubscriptionRequest& sr) {
+  // Check number of fields
+  unsigned long numFields;
+  if (rType == "EXECUTE")
```
There's inconsistent use of `{` here, but we recommend adding explicit scope `{}` for 1-line conditionals, it helps reduce future bugs.

theopolis commented on osquery/bro/utils.cpp:
```diff
@@ -0,0 +1,197 @@
+    sr.response_event = *broker::get<std::string>(msg[1]);
+  else {
+    return Status(1,
+                  "Unexpected data type; Response Event Name is not a string");
```
nitpick, "Response event name is not a string"

theopolis commented on osquery/bro/utils.cpp:
```diff
@@ -0,0 +1,197 @@
+      sr.response_topic = *broker::get<std::string>(msg[4]);
+    else {
+      return Status(
+          1, "Unexpected data type; Response Topic Name is not a string");
```
nitpick "Response topic name is not a string"

theopolis commented on osquery/bro/utils.cpp:
```diff
@@ -0,0 +1,197 @@
+  if (broker::is<std::string>(msg[1]))
+    sr.query = *broker::get<std::string>(msg[2]);
+  else {
+    return Status(1, "Unexpected data type; SQL query is not a string");
```
nitpick "SQL query is not a string"

theopolis commented on osquery/bro/utils.cpp:
```diff
@@ -0,0 +1,197 @@
+  if (broker::to_string(msg[4]).empty()) {
+    sr.response_topic = incoming_topic;
+    LOG(WARNING) << "No response topic given for event '" << sr.response_event
+                 << "'. Reporting back to "
```
Is it possible to collapse this warning into a single statement. We're a bit resistant to punctuation in log lines.

No response topic given for event 'EVENT' reporting back to incoming topic 'TOPIC'

theopolis commented on osquery/bro/utils.cpp:
```diff
@@ -0,0 +1,197 @@
+  // SUBSCRIBE or UNSUBSCRIBE
+  if (sr.snapshot) {
+    LOG(WARNING)
+        << "Only possible to query ADD and/or REMOVE for schedule queries";
```
nitpick "scheduled"

theopolis commented on osquery/bro/utils.cpp:
```diff
@@ -0,0 +1,197 @@
+    json_stream << clone;
+    pt::read_json(json_stream, groups_tree);
+
+    auto pt_groups = groups_tree.get_child("groups");
```
You should check that `groups` exists as a child of the tree, the `.count` method can be used.

Also, you could grab a `auto&`.

theopolis commented on osquery/bro/utils.cpp:
```diff
@@ -0,0 +1,197 @@
+
+    auto pt_groups = groups_tree.get_child("groups");
+    for (const auto& ptg : pt_groups) {
+      std::string ptg_value = pt_groups.get<std::string>(ptg.first);
```
This template cast version of `.get` can except. If you provide a second parameter as a default value, it will not:
```cpp
auto ptg_value = pt_groups.get<std::string>(ptg.first, "");
```

theopolis commented on osquery/dispatcher/bro.cpp:
```diff
@@ -0,0 +1,230 @@
+      sock = queue->fd();
+      if (FD_ISSET(sock, &fds)) {
+        // Process each message on this socket
+        for (auto& msg : queue->want_pop()) {
```
There's quite a bit of logic in this nested scope, from a style point of view, is it possible to refactor this out into another `static inline` method perhaps?

theopolis commented on osquery/dispatcher/bro.cpp:
```diff
@@ -0,0 +1,230 @@
+      FD_SET(sock, &fds); // each topic -> message_queue -> fd
+    }
+    // Wait for incoming message
+    if (select(sMax + 1, &fds, NULL, NULL, NULL) < 0) {
```
You can get better type safety and static analysis using `nullptr` instead of `0` for these pointers.

theopolis commented on osquery/dispatcher/bro.cpp:
```diff
@@ -0,0 +1,230 @@
+    FD_ZERO(&fds);
+    bm->getTopics(topics); // List of subscribed topics
+    int sMax = 0;
+    for (auto topic : topics) {
```
Can these be `const auto&`?

theopolis commented on osquery/dispatcher/bro.cpp:
```diff
@@ -0,0 +1,230 @@
+  // Wait for any requests
+  fd_set fds;
+  std::vector<std::string> topics;
+  int sock;
```
nitpick, mind initializing to `0`?
```cpp
int sock{0};
```

theopolis commented on osquery/dispatcher/bro.cpp:
```diff
@@ -0,0 +1,230 @@
+
+  // Wait for any requests
+  fd_set fds;
+  std::vector<std::string> topics;
```
You can reduce this scope into the `while`.

theopolis commented on osquery/dispatcher/bro.cpp:
```diff
@@ -0,0 +1,230 @@
+    }
+
+    // Check for the socket where a message arrived on
+    for (auto topic : topics) {
```
`auto&` or `const auto&`?

theopolis commented on osquery/dispatcher/bro.cpp:
```diff
@@ -0,0 +1,230 @@
+  // Set Broker groups and subscribe to group topics
+  std::vector<std::string> bro_groups;
+  parseBrokerGroups(FLAGS_bro_groups, bro_groups);
+  for (std::string g : bro_groups) {
```
nitpick, `for (const auto& group : bro_groups) {`

theopolis commented on osquery/dispatcher/bro.cpp:
```diff
@@ -0,0 +1,230 @@
+    }
+    // Wait for incoming message
+    if (select(sMax + 1, &fds, NULL, NULL, NULL) < 0) {
+      LOG(ERROR) << "Select returned an error code";
```
Would this be more helpful if the result from `select` was reported?

iBigQ commented:
Please have another look

theopolis commented:
ok to test

osqueryer commented:
:-1: The commit e9cc9e06496c6b358db5cc4253053a7a2c7f2eb6 (Job results: [423](https://jenkins.osquery.io/job/osqueryPullRequestOSX/423/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit e9cc9e06496c6b358db5cc4253053a7a2c7f2eb6 (Job results: [1132](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1132/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit e9cc9e06496c6b358db5cc4253053a7a2c7f2eb6 (Job results: [1134](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1134/)) failed one or more tests (**Windows**).

theopolis commented:
We could gate the build logic in CMake and the `provision.sh` script to make sure this only runs on Linux if that's the intention?

theopolis commented:
Is it possible to write some unit tests to make sure this is working, right now we at least know it compiles on Linux.

jedi22 commented on osquery/bro/BrokerManager.cpp:
```diff
@@ -0,0 +1,289 @@
+    nodeID_ = uid;
+    return Status(0, "OK");
+
+  } else {
```
You don't need this else, the return will suffice

jedi22 commented on osquery/bro/BrokerManager.cpp:
```diff
@@ -0,0 +1,289 @@
+Status BrokerManager::removeGroup(const std::string& group) {
+  auto element_pos = std::find(groups_.begin(), groups_.end(), group);
+  // Group exists?
+  if (element_pos != groups_.end()) {
```
```
if (element_pos == groups_.end()) {
    return Status(1, "Group '" + group + "' does not exist");
}
```
Then you can unindent the whole block

jedi22 commented on osquery/bro/BrokerManager.cpp:
```diff
@@ -0,0 +1,289 @@
+}
+
+Status BrokerManager::createMessageQueue(const std::string& topic) {
+  if (messageQueues_.count(topic) == 0) {
```
Same as above, flip the if so you can unindent the block.

jedi22 commented on osquery/bro/BrokerManager.cpp:
```diff
@@ -0,0 +1,289 @@
+  // Collect IPs
+  broker::vector addr_list;
+  SQL sql("SELECT address from interface_addresses");
+  if (sql.ok()) {
```
Flip if to remove indents

jedi22 commented on osquery/bro/BrokerManager.cpp:
```diff
@@ -0,0 +1,289 @@
+      }
+      const auto& value = row.at(colName);
+      switch (columnTypes.at(colName)) {
+      case ColumnType::UNKNOWN_TYPE: {
```
Cases don't need scoping braces but I'm not sure they hurt. @theopolis

jedi22 commented on osquery/bro/BrokerManager.cpp:
```diff
@@ -0,0 +1,289 @@
+      }
+      default: {
+        LOG(WARNING) << "Unknown ColumnType for column '" + colName + "'";
+        continue;
```
I think this continue is at the end of the loop block so is unneeded

jedi22 commented on osquery/bro/BrokerManager.h:
```diff
@@ -0,0 +1,93 @@
+  const std::string EVENT_HOST_UNSUBSCRIBE = "osquery::host_unsubscribe";
+
+  Status setNodeID(const std::string& uid);
+
```
I think this spacing is superfluous

iBigQ replied:
You mean the empty lines between two functions?

jedi22 replied:
Yeah, I think having them right up against each other is fine

jedi22 commented on osquery/bro/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+        BrokerManager.cpp
+        QueryManager.cpp
+        utils.cpp
+        )
```
Github is warning about the lack of newline here.

jedi22 commented on osquery/bro/QueryManager.cpp:
```diff
@@ -0,0 +1,168 @@
+  auto status = addQueryEntry(queryID, qr, "ONETIME");
+  if (status.ok()) {
+    return queryID;
+  } else {
```
Unneeded else statement

jedi22 commented on osquery/bro/QueryManager.cpp:
```diff
@@ -0,0 +1,168 @@
+  } else {
+    LOG(WARNING) << status.getMessage();
+  }
+  return "-1";
```
This is strange, will an empty string not convey the same message or is there a reason you can't do that?

jedi22 commented on osquery/bro/QueryManager.h:
```diff
@@ -0,0 +1,90 @@
+    return qm;
+  };
+
+  std::string addOneTimeQueryEntry(const SubscriptionRequest& qr);
```
Again, too much spacing

jedi22 commented on osquery/bro/utils.cpp:
```diff
@@ -0,0 +1,196 @@
+  unsigned long numFields;
+  if (rType == "EXECUTE") {
+    numFields = 6;
+  } else if (rType == "SUBSCRIBE") {
```
Crush these into an ||

jedi22 commented on osquery/bro/utils.cpp:
```diff
@@ -0,0 +1,196 @@
+
+  // Query String
+  if (broker::is<std::string>(msg[1])) {
+    sr.query = *broker::get<std::string>(msg[2]);
```
Is there anyway to do this not with pointers? (if no then ok, just making sure)

jedi22 commented on osquery/bro/utils.cpp:
```diff
@@ -0,0 +1,196 @@
+
+  // Response Event Name
+  if (broker::is<std::string>(msg[1])) {
+    sr.response_event = *broker::get<std::string>(msg[1]);
```
Same as above

jedi22 commented on osquery/bro/utils.cpp:
```diff
@@ -0,0 +1,196 @@
+                 << "' reporting back to "
+                    "incoming topic '"
+                 << incoming_topic << "'";
+  } else {
```
Crush into else if, else

jedi22 commented on osquery/bro/utils.h:
```diff
@@ -0,0 +1,40 @@
+Status parseBrokerGroups(const std::string& json_groups,
+                         std::vector<std::string>& groups);
+
+Status printQueryLogItem(const QueryLogItem& item);
```
Spacing

jedi22 commented on osquery/dispatcher/bro.cpp:
```diff
@@ -0,0 +1,227 @@
+  // Set Broker UID
+  std::string ident;
+  auto status_huuid = getHostUUID(ident);
+  if (status_huuid.ok())
```
Use braces even for one line ifs

iBigQ commented:
@theopolis is there a more detailed jenkins log output to see what actually went wrong?

In a last resort we can compile it on Linux only.

osqueryer commented:
:-1: The commit aec065a7d40481998b968ee354ce2d62fc92e7fe (Job results: [431](https://jenkins.osquery.io/job/osqueryPullRequestOSX/431/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit 48bbd07c8b90709cf2ec10f6bac0fa57b4882cf7 (Job results: [432](https://jenkins.osquery.io/job/osqueryPullRequestOSX/432/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit e61efa4bfbd88fe050bba3d6f22536f55035a75e (Job results: [433](https://jenkins.osquery.io/job/osqueryPullRequestOSX/433/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit e61efa4bfbd88fe050bba3d6f22536f55035a75e (Job results: [1144](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1144/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit e61efa4bfbd88fe050bba3d6f22536f55035a75e (Job results: [4203](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4203/)) failed one or more tests (**Linux**).

iBigQ commented:
Here are the TODOs:

- [ ] Unit tests
- [ ] Cross platform support (or linux only)

I guess the second depends solely on broker (communication library)

osqueryer commented:
:-1: The commit e61efa4bfbd88fe050bba3d6f22536f55035a75e (Job results: [4204](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4204/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 3c2319d5ac99a384f0358da47aa16e45b9c9a9f3 (Job results: [4208](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4208/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 3c2319d5ac99a384f0358da47aa16e45b9c9a9f3 (Job results: [1147](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1147/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 3c2319d5ac99a384f0358da47aa16e45b9c9a9f3 (Job results: [4209](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4209/)) failed one or more tests (**Linux**).

iBigQ commented:
@theopolis jenkins does not make new deps. But this is what I need since I pushed an update to a formula. It now runs on mac but I also want to test the new version on linux as well.

Related, can you upload the new boost bottle to make the centos7 test not to fail when compiling boost from source?

I would also like to see if caf/broker compiles on windows. You use choco as far as I see. Can you bootstrap me into what needs to be done to create a package for/within the osquery project?

Thanks :)

poppyseedplehzr commented:
ok to test

osqueryer commented:
:no_good: The commit 3c2319d5ac99a384f0358da47aa16e45b9c9a9f3 (Job results: [1137](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1137/)) failed the code audit or documentation test.

osqueryer commented:
:-1: The commit 3c2319d5ac99a384f0358da47aa16e45b9c9a9f3 (Job results: [4210](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4210/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 3c2319d5ac99a384f0358da47aa16e45b9c9a9f3 (Job results: [1148](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1148/)) failed one or more tests (**Windows**).

poppyseedplehzr commented:
ok to test

osqueryer commented:
:no_good: The commit 3c2319d5ac99a384f0358da47aa16e45b9c9a9f3 (Job results: [1138](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1138/)) failed the code audit or documentation test.

osqueryer commented:
:-1: The commit 3c2319d5ac99a384f0358da47aa16e45b9c9a9f3 (Job results: [1149](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1149/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 3c2319d5ac99a384f0358da47aa16e45b9c9a9f3 (Job results: [4212](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4212/)) failed one or more tests (**Linux**).

theopolis commented:
@iBigQ for the `choco` integration we need the `broker` application/libraries to be available in the choco repo, or we maintain our own development package. I'll defer to @poppyseedplehzr for details.

poppyseedplehzr commented:
Hey @iBigQ, I built us a caf package on `0.14.5`, but wasn't _quite_ able to get broker fully building. If you do the following to our `provision.ps1` script:
```
C:\Users\Nicholas\projects\repos\osquery [master ≡ +0 ~1 -0 !]> git diff
diff --git a/tools/provision.ps1 b/tools/provision.ps1
index 3ea99459..067e31f0 100644
--- a/tools/provision.ps1
+++ b/tools/provision.ps1
@@ -270,6 +270,7 @@ function Install-ThirdParty {
   $packages = @(
     "boost-msvc14.1.63.0",
     "bzip2.1.0.6",
+    "caf.0.14.5",
     "doxygen.1.8.11",
     "gflags-dev.2.2.0",
     "glog.0.3.4",
```
Then run `.\tools\make-win64-dev-env.bat` to install `caf`, and finally run through something similar to this:
```
git clone https://github.com/bro/broker.git
cd broker
git submodule update --recursive --init
# Apply this patch:

C:\Users\Nicholas\Downloads\broker\cmake [(d29fbf6...) +0 ~1 -0 !]> git diff
diff --git a/FindCAF.cmake b/FindCAF.cmake
index ea2860c..3259b38 100644
--- a/FindCAF.cmake
+++ b/FindCAF.cmake
@@ -43,7 +43,8 @@ foreach (comp ${CAF_FIND_COMPONENTS})
               /usr/local/include
               /opt/local/include
               /sw/include
-              ${CMAKE_INSTALL_PREFIX}/include)
+              ${CMAKE_INSTALL_PREFIX}/include
+              $ENV{CAF_HEADER_LOCATION})
   mark_as_advanced(CAF_INCLUDE_DIR_${UPPERCOMP})
   if (NOT "${CAF_INCLUDE_DIR_${UPPERCOMP}}"
       STREQUAL "CAF_INCLUDE_DIR_${UPPERCOMP}-NOTFOUND")
@@ -74,7 +75,8 @@ foreach (comp ${CAF_FIND_COMPONENTS})
                      /usr/local/lib
                      /opt/local/lib
                      /sw/lib
-                     ${CMAKE_INSTALL_PREFIX}/lib)
+                     ${CMAKE_INSTALL_PREFIX}/lib
+                     $ENV{CAF_LIB_LOCATION})
       mark_as_advanced(CAF_LIBRARY_${UPPERCOMP})
       if ("${CAF_LIBRARY_${UPPERCOMP}}" STREQUAL "CAF_LIBRARY-NOTFOUND")
         set(CAF_${comp}_FOUND false)

$env:CAF_HEADER_LOCATION="C:\ProgramData\chocolatey\lib\caf\local\include"
$env:CAF_LIB_LOCATION="C:\ProgramData\chocolatey\lib\caf\local\lib"

# Then this patch...

C:\Users\Nicholas\Downloads\broker [master ≡ +1 ~2 -0 !]> git diff CMakeLists.txt
diff --git a/CMakeLists.txt b/CMakeLists.txt
index e439cde..25d0d16 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -161,7 +161,7 @@ endif ()
 add_subdirectory(bindings)

 enable_testing()
-add_subdirectory(tests)
+#add_subdirectory(tests)

 string(TOUPPER ${CMAKE_BUILD_TYPE} BuildType)

@@ -190,7 +190,7 @@ display(ENABLE_SHARED yes shared_summary)
 display(ENABLE_STATIC yes static_summary)
 display(CAF_FOUND "${caf_dir} (${CAF_VERSION})" caf_summary)
 display(ROCKSDB_FOUND ${ROCKSDB_INCLUDE_DIRS} rocksdb_summary)
-display(BROKER_PYTHON_BINDINGS ${BROKER_PYTHON_BINDINGS} pybindings_summary)
+#display(BROKER_PYTHON_BINDINGS ${BROKER_PYTHON_BINDINGS} pybindings_summary)

 message(
     "\n==================|  broker Config Summary  |===================="

# Then build your solution.
mkdir win-build
cd win-build
cmake -DROCKSDB_LIBRARIES=C:\ProgramData\chocolatey\lib\rocksdb\local\lib -DROCKSDB_INCLUDE_DIRS=C:\ProgramData\chocolatey\lib\rocksdb\local\include -DHAVE_CXX11=yes -DHAVE_BROKER_BIG_ENDIAN=yes ../ -G "Visual Studio 14 2015 Win64"
start -D broker.sln

# Open up broker properties -> C/C++ -> Command Line -> Remove the -Wno-unused.
```

You can get _really_ close, but the builds were failing for me with numerous errors. Let me know if you have any more luck. I'll give this another shot later but I have a few other choco packages that need building. if you can get a solid set of steps to repro the broker lib builds let me know and I'll be more than happy to throw a chocolatey package up for you :)

If you're interested in rolling your own choco packages, you can check out your `C:\ProgramData\chocolatey\lib\caf\` directory and get a feel for what the directory structure looks like. Honestly you just re-create that dir structure and run `choco pack` at the root, in the same folder as the `.nuspec`, and you'll have a chocolatey package.

Hope this helps!

osqueryer commented:
:-1: The commit 7061d801bce04feda6cb38c706ba12c07e98662e (Job results: [1228](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1228/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit ba98e94b3f1a1a7413e2089c0da9585f47498fd8 (Job results: [1326](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1326/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit f1de06e92d0cb31a27a1c71f386d5ca74beb8712 (Job results: [899](https://jenkins.osquery.io/job/osqueryPullRequestOSX/899/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit f1de06e92d0cb31a27a1c71f386d5ca74beb8712 (Job results: [900](https://jenkins.osquery.io/job/osqueryPullRequestOSX/900/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:no_good: The commit f1de06e92d0cb31a27a1c71f386d5ca74beb8712 (Job results: [1580](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1580/)) failed the code audit or documentation test.

osqueryer commented:
:-1: The commit f1de06e92d0cb31a27a1c71f386d5ca74beb8712 (Job results: [4758](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4758/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit f1de06e92d0cb31a27a1c71f386d5ca74beb8712 (Job results: [4759](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4759/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit f1de06e92d0cb31a27a1c71f386d5ca74beb8712 (Job results: [1629](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1629/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit f1de06e92d0cb31a27a1c71f386d5ca74beb8712 (Job results: [1630](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1630/)) failed one or more tests (**Windows**).

osqueryer commented:
:no_good: The commit 0ebcd3ea9faf3830c98c91e8bfd910ac3e041220 (Job results: [1615](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1615/)) failed the code audit or documentation test.

osqueryer commented:
:-1: The commit 0ebcd3ea9faf3830c98c91e8bfd910ac3e041220 (Job results: [1671](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1671/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 0ebcd3ea9faf3830c98c91e8bfd910ac3e041220 (Job results: [1672](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1672/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 0ebcd3ea9faf3830c98c91e8bfd910ac3e041220 (Job results: [4799](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4799/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 0ebcd3ea9faf3830c98c91e8bfd910ac3e041220 (Job results: [4800](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4800/)) failed one or more tests (**Linux**).

iBigQ commented:
@theopolis Any idea why ThinLTO fails on compiling this PR?

```
[ 81%] Linking CXX executable osquery_tables_tests
cd /home/osquery/osquery/build/xenial/osquery && /usr/local/osquery/Cellar/cmake/3.6.1_100/bin/cmake -E cmake_link_script CMakeFiles/osquery_tables_tests.dir/link.txt --verbose=1
/usr/local/osquery/bin/ccache /usr/local/osquery/bin/clang++   -L/usr/local/osquery/legacy/lib -L/usr/local/osquery/lib    -flto=thin CMakeFiles/osquery_tables_tests.dir/main/tests.cpp.o CMakeFiles/osquery_tables_tests.dir/tables/events/tests/file_events_tests.cpp.o CMakeFiles/osquery_tables_tests.dir/tables/networking/tests/networking_tables_tests.cpp.o CMakeFiles/osquery_tables_tests.dir/tables/networking/linux/tests/iptables_tests.cpp.o CMakeFiles/osquery_tables_tests.dir/tables/system/tests/system_tables_tests.cpp.o CMakeFiles/osquery_tables_tests.dir/tables/system/linux/tests/portage_tests.cpp.o CMakeFiles/osquery_tables_tests.dir/tables/system/linux/tests/users_tests.cpp.o CMakeFiles/osquery_tables_tests.dir/tables/yara/tests/yara_tests.cpp.o  -o osquery_tables_tests -rdynamic -Wl,-whole-archive libosquery.a -Wl,-no-whole-archive -Wl,-whole-archive libosquery_additional.a -Wl,-no-whole-archive ../third-party/googletest/googlemock/gtest/libgtest.a ../third-party/googletest/googlemock/libgmock.a libosquery_testing.a -lz -Wl,-Bstatic -lgflags -lthrift -lglog -lbroker -lcaf_core_static -lcaf_io_static -Wl,-Bdynamic -ldl -Wl,-Bstatic -lboost_system-mt -lboost_filesystem-mt -lboost_context-mt -Wl,-Bdynamic -lrt -lc -rdynamic -Wl,-zrelro -Wl,-znow -pie -lgcc_s -Wl,--build-id -fuse-ld=lld -static-libstdc++ -Wl,-Bstatic -ludev -laudit -laws-cpp-sdk-kinesis -laws-cpp-sdk-firehose -laws-cpp-sdk-sts -laws-cpp-sdk-core -larchive -Wl,-Bdynamic -lresolv -Wl,-Bstatic -lcryptsetup -ldevmapper -lgcrypt -lgpg-error -lblkid -lip4tc -lapt-pkg -ldpkg -llz4 -llzma -lrpm -lrpmio -lbeecrypt -lpopt -ldb -laugeas -lfa -lxml2 -ltsk -lyara -llldpctl -lrocksdb_lite -lboost_regex-mt -lcppnetlib-uri -lcppnetlib-client-connections -lsnappy -lssl -lcrypto -Wl,-Bdynamic -lpthread -Wl,-Bstatic -llinenoise -lmagic -lbz2 -luuid -Wl,-Bdynamic -lpthread 
/usr/local/osquery/bin/ld.lld: error: Expected at most one ThinLTO module per bitcode file
clang-4.0: error: linker command failed with exit code 1 (use -v to see invocation)
osquery/CMakeFiles/osquery_tables_tests.dir/build.make:336: recipe for target 'osquery/osquery_tables_tests' failed
make[3]: *** [osquery/osquery_tables_tests] Error 1
CMakeFiles/Makefile2:633: recipe for target 'osquery/CMakeFiles/osquery_tables_tests.dir/all' failed
make[2]: *** [osquery/CMakeFiles/osquery_tables_tests.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make[1]: *** [all] Error 2
Makefile:53: recipe for target 'all' failed
make: *** [all] Error 2
```

osqueryer commented:
:no_good: The commit 4f837f39f9189fa46b952191766217ae8afe7242 (Job results: [1616](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1616/)) failed the code audit or documentation test.

osqueryer commented:
:no_good: The commit 4adddaac8133ae934b7d90940111d7267380e9c3 (Job results: [1617](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1617/)) failed the code audit or documentation test.

osqueryer commented:
:-1: The commit 4adddaac8133ae934b7d90940111d7267380e9c3 (Job results: [1674](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1674/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 4adddaac8133ae934b7d90940111d7267380e9c3 (Job results: [1675](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1675/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 4adddaac8133ae934b7d90940111d7267380e9c3 (Job results: [4802](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4802/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 4adddaac8133ae934b7d90940111d7267380e9c3 (Job results: [4803](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4803/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 4d96cc74ba154cb215684f2466da97e0c0856d5e (Job results: [1078](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1078/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit 4d96cc74ba154cb215684f2466da97e0c0856d5e (Job results: [4937](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4937/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 4d96cc74ba154cb215684f2466da97e0c0856d5e (Job results: [1079](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1079/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:no_good: The commit 4d96cc74ba154cb215684f2466da97e0c0856d5e (Job results: [1741](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1741/)) failed the code audit or documentation test.

osqueryer commented:
:-1: The commit 4d96cc74ba154cb215684f2466da97e0c0856d5e (Job results: [4938](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4938/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 4d96cc74ba154cb215684f2466da97e0c0856d5e (Job results: [1815](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1815/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 4d96cc74ba154cb215684f2466da97e0c0856d5e (Job results: [1816](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1816/)) failed one or more tests (**Windows**).

osqueryer commented:
:no_good: The commit da01bb799849b395933fcacd455ca9f239f6e9ee (Job results: [1742](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1742/)) failed the code audit or documentation test.

osqueryer commented:
:-1: The commit da01bb799849b395933fcacd455ca9f239f6e9ee (Job results: [1817](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1817/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit da01bb799849b395933fcacd455ca9f239f6e9ee (Job results: [1818](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1818/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 6e987710b641671765c0e1071fba0d06a20701d5 (Job results: [1081](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1081/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit 6e987710b641671765c0e1071fba0d06a20701d5 (Job results: [1082](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1082/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:no_good: The commit 6e987710b641671765c0e1071fba0d06a20701d5 (Job results: [1743](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1743/)) failed the code audit or documentation test.

osqueryer commented:
:-1: The commit 6e987710b641671765c0e1071fba0d06a20701d5 (Job results: [4940](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4940/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 6e987710b641671765c0e1071fba0d06a20701d5 (Job results: [4941](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4941/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 6e987710b641671765c0e1071fba0d06a20701d5 (Job results: [1819](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1819/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 6e987710b641671765c0e1071fba0d06a20701d5 (Job results: [1820](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1820/)) failed one or more tests (**Windows**).

osqueryer commented:
:no_good: The commit 4c2f38e58fc885c167fd431ab0ae742bdd3b51a6 (Job results: [1767](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1767/)) failed the code audit or documentation test.

osqueryer commented:
:-1: The commit e6d3a7e688e0640be4ce7cfa4b1fec39504be390 (Job results: [1853](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1853/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit e6d3a7e688e0640be4ce7cfa4b1fec39504be390 (Job results: [1854](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1854/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 218ed9f8dfb1cb130b2356b691b94aa43017198e (Job results: [4970](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4970/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 218ed9f8dfb1cb130b2356b691b94aa43017198e (Job results: [1856](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1856/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 218ed9f8dfb1cb130b2356b691b94aa43017198e (Job results: [1857](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1857/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 8d5d649c65b6d4f218ea7cea8b2ec500544b86cf (Job results: [4972](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4972/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 8d5d649c65b6d4f218ea7cea8b2ec500544b86cf (Job results: [1858](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1858/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 8d5d649c65b6d4f218ea7cea8b2ec500544b86cf (Job results: [4973](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4973/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 8d5d649c65b6d4f218ea7cea8b2ec500544b86cf (Job results: [1859](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1859/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 46ae554b3fc975eddaa629100a516957944eea11 (Job results: [4982](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4982/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 46ae554b3fc975eddaa629100a516957944eea11 (Job results: [1867](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1867/)) failed one or more tests (**Windows**).

iBigQ commented:
Current limitations:

- [x] ThinLTO disabled
- [x] Skip bro_integration on Windows

osqueryer commented:
:-1: The commit 4b480943dfcd7ec0308b0a53d39436535f29cce6 (Job results: [4984](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4984/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 4b480943dfcd7ec0308b0a53d39436535f29cce6 (Job results: [1869](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1869/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 4b480943dfcd7ec0308b0a53d39436535f29cce6 (Job results: [4985](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4985/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 4b480943dfcd7ec0308b0a53d39436535f29cce6 (Job results: [1870](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1870/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 0042715b9bd1884143a236cc162a0d0cabb8188f (Job results: [1871](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1871/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 8a53f5b3b24e5c0ae616ab81171803d820ec61f6 (Job results: [1886](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1886/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 8a53f5b3b24e5c0ae616ab81171803d820ec61f6 (Job results: [1887](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1887/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 32b8a3168501374c211c3741acd25d6caa92336b (Job results: [5003](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5003/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit fde6a9d1d614db4e383e40db28bd7fde09b20537 (Job results: [5005](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5005/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit fde6a9d1d614db4e383e40db28bd7fde09b20537 (Job results: [5006](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5006/)) failed one or more tests (**Linux**).

iBigQ commented:
@theopolis @poppyseedplehzr @jedi22 
I addressed your feedback and continued development. Here is a brief summary:

- Compiles on Linux and Mac
  - No idea why some tests fail during the CI check, does not seem to be related
  - Cmake skips bro support on Windows
  - Tested in a demo setup (including Bro instance) on Linux
- Bro-Osquery runs as a distributed plugin
- Added tests for read/write messages (BrokerManager) and query handling (QueryManager)

Ready for review

osqueryer commented:
:-1: The commit da237548fcb9b194287f31056c7c169864ebf6a8 (Job results: [1158](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1158/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit da237548fcb9b194287f31056c7c169864ebf6a8 (Job results: [5032](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5032/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit da237548fcb9b194287f31056c7c169864ebf6a8 (Job results: [1159](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1159/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit 25f056e505b89a057e916f14155ded347858fa19 (Job results: [5045](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5045/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 25f056e505b89a057e916f14155ded347858fa19 (Job results: [5046](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5046/)) failed one or more tests (**Linux**).

theopolis commented on osquery/distributed/CMakeLists.txt:
```diff
@@ -2,9 +2,15 @@ ADD_OSQUERY_LIBRARY(TRUE osquery_distributed
 )
 
+if(NOT WINDOWS)
+    LIST(APPEND OSQUERY_DISTRIBUTED_PLUGINS "plugins/bro_distributed.cpp")
```
nit, 3 spaces, should be 2

theopolis commented on osquery/distributed/plugins/bro_distributed.cpp:
```diff
@@ -0,0 +1,336 @@
+  BrokerManager& bm = BrokerManager::get();
+
+  // Subscribe to all and individual topic
+  Status s_mq_all = bm.createMessageQueue(bm.TOPIC_ALL);
```
If you're not using these `Status` objects simultaneously you can reuse/copy-assign the same local variable.

```cpp
auto s =  bm.createMessageQueue(bm.TOPIC_ALL);
if (!s.ok()) {
  ...
}

s = bm.createMessageQueue(bm.TOPIC_PRE_INDIVIDUALS + bm.getNodeID());
....
```

theopolis commented on osquery/distributed/plugins/bro_distributed.cpp:
```diff
@@ -0,0 +1,336 @@
+  QueryManager& qm = QueryManager::get();
+
+  // Check Event Type
+  if (msg.size() < 1 or !broker::is<std::string>(msg[0])) {
```
nit, `s/or/||/`

theopolis commented on osquery/distributed/plugins/bro_distributed.cpp:
```diff
@@ -0,0 +1,336 @@
+      return s_unsub;
+    }
+
+  } else if (eventName == "osquery::host_test") {
```
what?

theopolis commented on osquery/distributed/plugins/bro_distributed.cpp:
```diff
@@ -0,0 +1,336 @@
+  } else if (eventName == "osquery::host_test") {
+  } else {
+    // Unkown Message
+    return Status(7, "Unknown event name '" + eventName + "'");
```
Hmm, if you're using some `Status` values other than `0` and `1`, do you mind creating a:

```cpp
enum class BroStatuses {
  BroErrorSomething = 7,
  ...
};
```
So we can keep track of what these mean? Or you can just have a set of `#define BRO_ERROR_SOMETHING 7`.

iBigQ replied:
What is the projects practice on this? I dont care to use 0/1 only. I introduced different error numbers to help in debugging and finding the root cause if a functions can potentially fail for different reasons. These codes are therefore not global but function-specific.
Enum about Bro Statuses might be overkilled, I will limit all error codes to 1.

theopolis commented on osquery/distributed/plugins/bro_distributed.cpp:
```diff
@@ -0,0 +1,336 @@
+  // Apply to new config/schedule
+  std::map<std::string, std::string> config_schedule;
+  config_schedule["bro"] = qm.getQueryConfigString();
+  VLOG(1) << "Applying new schedule: " << config_schedule["bro"];
```
Note, we don't use verbose log levels anywhere, and outputting a bunch of JSON in verbose modes may be removed in the future.

theopolis commented on osquery/distributed/plugins/bro_distributed.cpp:
```diff
@@ -0,0 +1,336 @@
+  sMax = std::max(sMax, bm.getOutgoingConnectionFD());
+
+  // Wait for incoming message
+  int select_code = select(sMax + 1, &fds, nullptr, nullptr, nullptr);
```
Use `poll` instead, we had a nasty issue with `select` as we recommend increasing FD max past 1024, and make use of `inotify` persistent FDs so you may overflow your `fd_set`.

theopolis commented:
I will try to dig into the ThinLTO errors with caf and broker today. :)

osqueryer commented:
:no_good: The commit ff964ee3fdbbf4cff9a3945a8f1fc6d26544214f (Job results: [1881](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1881/)) failed the code audit or documentation test.

osqueryer commented:
:-1: The commit 77f72bf1d6a1ca3d63492003371c8fdd5f11e86c (Job results: [125](https://jenkins.osquery.io/job/osqueryPullRequestLinux/125/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:-1: The commit 77f72bf1d6a1ca3d63492003371c8fdd5f11e86c (Job results: [126](https://jenkins.osquery.io/job/osqueryPullRequestLinux/126/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:no_good: The commit c561cdac18dd426224fa4dbb1fcf509438ba025b (Job results: [1894](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1894/)) failed the code audit or documentation test.

theopolis commented on tools/provision/formula/broker.rb:
```diff
@@ -0,0 +1,100 @@
+  desc "Broker Communication Library"
+  homepage "https://github.com/bro/broker"
+  url "https://github.com/bro/broker.git",
+	  :tag => "v0.6" # Need git url for recursive clone
```
This should use a commit hash so the tag cannot be swapped out under our noses.

theopolis commented:
@iBigQ, this should work with ThinLTO now. The error is hard to debug, but it's a result of the `lld` linker having two implementation files with the same base filename, haha!

I'll ask @ryanheffernan, @poppyseedplehzr, @jedi22, or anyone else who spends a bit of time in our code base to do a few more reviews as this gets cleaned up and polished.

iBigQ commented:
FreeBSD fails because the new dependencies (formulas) are not installed. Requires `make deps` on the CI build host

iBigQ commented:
@theopolis thanks for finding the reason for the ThinLTO error!

osqueryer commented:
:-1: The commit ec5cbedd05dc527a1f0d3c6bf4fd2bede37759dc (Job results: [1252](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1252/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit ec5cbedd05dc527a1f0d3c6bf4fd2bede37759dc (Job results: [5132](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5132/)) failed one or more tests (**Linux**).

iBigQ commented:
@theopolis Seems I messed up the history. You already squashed my commits? I was not aware of and now it still contains all my original commits (e.g. also disabling ThinLTO). 
Since I am not sure what you did, can you fix it/roll back to what you did. I can even redo my latest changes (since your squashing)

theopolis commented:
>  Seems I messed up the history. 
Nope, I messed it up by squashing.

> Since I am not sure what you did, can you fix it/roll back to what you did. I can even redo my latest changes (since your squashing)
I can just go from where it's at now.

I squashed all of your commits into one because it's easier to work with. :) 60+ commits in a PR is getting a little out of hand, haha.

theopolis commented:
Do you mind going through and adding the right docblocks for classes and their methods/members? As well as removing extraneous comments and TODOs? If you want to merge with TODOs we'll need to have issues tracking them.

osqueryer commented:
:-1: The commit dfc07da1815d72b5d8d281f214dcc537c5aad5ed (Job results: [1351](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1351/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit dfc07da1815d72b5d8d281f214dcc537c5aad5ed (Job results: [5235](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5235/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit dfc07da1815d72b5d8d281f214dcc537c5aad5ed (Job results: [5236](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5236/)) failed one or more tests (**Linux**).

iBigQ commented:
From linux CI: `/usr/local/osquery/bin/ld.lld: error: failed to open osquery_additional_tests: No space left on device`. This is related to the testing infrastructure only, isnt it?

osqueryer commented:
:-1: The commit 2b64dd331e79c49f53840f7909061dde6d6dae13 (Job results: [5245](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5245/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 2b64dd331e79c49f53840f7909061dde6d6dae13 (Job results: [5246](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5246/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 1c5c7511967ba465393b379a6cafa6afab39c434 (Job results: [5261](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5261/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 1c5c7511967ba465393b379a6cafa6afab39c434 (Job results: [5262](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5262/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 1c5c7511967ba465393b379a6cafa6afab39c434 (Job results: [263](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/263/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:-1: The commit 1c5c7511967ba465393b379a6cafa6afab39c434 (Job results: [264](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/264/)) failed one or more tests (**FreeBSD**).

theopolis commented:
Ok the plan is to bring this into master in the next release. Need to make another pass at why Linux and FreeBSD are failing.

osqueryer commented:
:no_good: The commit ecc5e6b5db1e49cfc5194470a35001b44a172d5a (Job results: [2237](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2237/)) failed the code audit or documentation test.

osqueryer commented:
:-1: The commit ecc5e6b5db1e49cfc5194470a35001b44a172d5a (Job results: [1632](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1632/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit ecc5e6b5db1e49cfc5194470a35001b44a172d5a (Job results: [5522](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5522/)) failed one or more tests (**Linux**).

osqueryer commented:
:no_good: The commit 565f29de5819cd1e1d73be4ebb834c3f6744a300 (Job results: [2238](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2238/)) failed the code audit or documentation test.

osqueryer commented:
:-1: The commit 565f29de5819cd1e1d73be4ebb834c3f6744a300 (Job results: [5524](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5524/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 565f29de5819cd1e1d73be4ebb834c3f6744a300 (Job results: [5525](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5525/)) failed one or more tests (**Linux**).

osqueryer commented:
:no_good: The commit fe196c50e3750745896185421b6442ff41a05bdb (Job results: [2239](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2239/)) failed the code audit or documentation test.

osqueryer commented:
:-1: The commit fe196c50e3750745896185421b6442ff41a05bdb (Job results: [5526](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5526/)) failed one or more tests (**Linux**).

iBigQ commented:
@theopolis Please check out PR #3559 to fix the Code Audit

osqueryer commented:
:-1: The commit 7126cad37fab5598b39ba892d8ef57aa6e55413e (Job results: [5550](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5550/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 7126cad37fab5598b39ba892d8ef57aa6e55413e (Job results: [5551](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5551/)) failed one or more tests (**Linux**).

theopolis commented:
I would say, take a pass at the status messages. We recommend 1-sentence log lines without punctuation. Double check for commented-out code lines too.

osqueryer commented:
:-1: The commit 3569dbfa82e1262fbe5746e871bf9e6c66a1e8cc (Job results: [619](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/619/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:-1: The commit 3569dbfa82e1262fbe5746e871bf9e6c66a1e8cc (Job results: [1788](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1788/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit 3569dbfa82e1262fbe5746e871bf9e6c66a1e8cc (Job results: [1789](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1789/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:no_good: The commit 3569dbfa82e1262fbe5746e871bf9e6c66a1e8cc (Job results: [2360](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2360/)) failed the code audit or documentation test.

osqueryer commented:
:-1: The commit 3569dbfa82e1262fbe5746e871bf9e6c66a1e8cc (Job results: [620](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/620/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:-1: The commit 3569dbfa82e1262fbe5746e871bf9e6c66a1e8cc (Job results: [5682](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5682/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 3569dbfa82e1262fbe5746e871bf9e6c66a1e8cc (Job results: [5683](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5683/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 3569dbfa82e1262fbe5746e871bf9e6c66a1e8cc (Job results: [2562](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2562/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 3569dbfa82e1262fbe5746e871bf9e6c66a1e8cc (Job results: [2563](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2563/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 9891946a56d19ecb9b9d6b4f147e4795983ffb1e (Job results: [1792](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1792/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit 9891946a56d19ecb9b9d6b4f147e4795983ffb1e (Job results: [5686](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5686/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 9891946a56d19ecb9b9d6b4f147e4795983ffb1e (Job results: [1793](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1793/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit 9891946a56d19ecb9b9d6b4f147e4795983ffb1e (Job results: [5687](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5687/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 9891946a56d19ecb9b9d6b4f147e4795983ffb1e (Job results: [2566](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2566/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit bc254f835f35fa1d78d7da1892f19f6fb1025d7e (Job results: [1795](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1795/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit bc254f835f35fa1d78d7da1892f19f6fb1025d7e (Job results: [5689](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5689/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit bc254f835f35fa1d78d7da1892f19f6fb1025d7e (Job results: [5690](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5690/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit bc254f835f35fa1d78d7da1892f19f6fb1025d7e (Job results: [674](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/674/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:-1: The commit bc254f835f35fa1d78d7da1892f19f6fb1025d7e (Job results: [1846](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1846/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit bc254f835f35fa1d78d7da1892f19f6fb1025d7e (Job results: [5743](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5743/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit bc254f835f35fa1d78d7da1892f19f6fb1025d7e (Job results: [2629](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2629/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit bc254f835f35fa1d78d7da1892f19f6fb1025d7e (Job results: [5746](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5746/)) failed one or more tests (**Linux**).
